zebrad/src/components/mempool/downloads.rs

```diff
@@ -213,19 +224,39 @@ where
 
         let network = self.network.clone();
         let verifier = self.verifier.clone();
-        let state = self.state.clone();
+        let mut state = self.state.clone();
+        let mempool = self.mempool.clone();
 
         let fut = async move {
-            // TODO: adapt this for transaction / mempool
-            // // Check if the block is already in the state.
-            // // BUG: check if the hash is in any chain (#862).
-            // // Depth only checks the main chain.
-            // match state.oneshot(zs::Request::Depth(hash)).await {
-            //     Ok(zs::Response::Depth(None)) => Ok(()),
-            //     Ok(zs::Response::Depth(Some(_))) => Err("already present".into()),
-            //     Ok(_) => unreachable!("wrong response"),
-            //     Err(e) => Err(e),
-            // }?;
+            // Check if the transaction is already in the state.
+            match state
+                .ready_and()
+                .await?
+                .call(zs::Request::Transaction(txid.mined_id()))
+                .await
+            {
+                Ok(zs::Response::Transaction(None)) => Ok(()),
+                Ok(zs::Response::Transaction(Some(_))) => Err("already present in state".into()),
+                Ok(_) => unreachable!("wrong response"),
+                Err(e) => Err(e),
+            }?;
+            // Check if the transaction is already in the mempool.
```

Comment:
Performance nitpick: We might want to do the mempool check first.

The mempool is always in memory, so the check will be faster.

The transaction ID index could be on disk, so the check will be slower, so we only want to do it if we have to.
(If the transaction is not in the mempool.)